staging/src/k8s.io/component-base/metrics/prometheusextension/timing_histogram_test.go

```diff
@@ -0,0 +1,125 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package prometheusextension
+
+import (
+	"math"
+	"testing"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	dto "github.com/prometheus/client_model/go"
+
+	testclock "k8s.io/utils/clock/testing"
+)
+
+func TestTimingHistogramNonMonotonicBuckets(t *testing.T) {
+	testCases := map[string][]float64{
+		"not strictly monotonic":  {1, 2, 2, 3},
+		"not monotonic at all":    {1, 2, 4, 3, 5},
+		"have +Inf in the middle": {1, 2, math.Inf(+1), 3},
+	}
+	for name, buckets := range testCases {
+		_, err := NewTimingHistogram(TimingHistogramOpts{
+			Name:    "test_histogram",
+			Help:    "helpless",
+			Buckets: buckets,
+		})
+		if err == nil {
+			t.Errorf("Buckets %v are %s but NewHTimingistogram did not complain.", buckets, name)
+		}
+	}
+}
+
+var testBuckets = []float64{0, 0.5, 1}
+var value0 float64 = 0.25
+
+func exerciseTimingHistogram(th TimingHistogram, t0 time.Time, clk *testclock.FakePassiveClock) func(t *testing.T) {
+	return func(t *testing.T) {
+		v0 := value0
+		t1 := t0.Add(time.Nanosecond)
+		var v1 float64 = 0.75
+		clk.SetTime(t1)
+		th.Set(v1)
+		t2 := t1.Add(time.Microsecond)
+		var d2 float64 = 0.5
+		v2 := v1 + d2
+		clk.SetTime(t2)
+		th.Add(d2)
+		t3 := t2
+		for i := 0; i < 1000000; i++ {
+			t3 = t3.Add(time.Nanosecond)
+			clk.SetTime(t3)
+			th.Set(v2)
+		}
+		var d3 float64 = -0.6
+		v3 := v2 + d3
+		th.Add(d3)
+		t4 := t3.Add(time.Second)
+		clk.SetTime(t4)
+
+		metch := make(chan prometheus.Metric)
```

Comment:
this test setup is really hard to read and it's not obvious to me what you are doing...

Reply:
In my latest push I added a func comment hopefully explaining it.